Implement SdJwtVP.of(String) with enhanced error handling

Ogenbertrand opened the pull request:
This PR implements the SdJwtVP.of(String) method with improved error handling. It adds validation to handle missing disclosures, and invalid disclosure data during the disclosure processing flow.

Changes Implemented
Added a check for missing disclosures, which throws an IllegalArgumentException if none are found.
Improved error handling for invalid or malformed disclosure data by throwing appropriate exceptions.

Testing
Unit tests have been added to cover the following cases:
Missing disclosures
Invalid/malformed disclosure data

Issue Reference
Closes #33020

Ogenbertrand committed:
Implement SdJwtVP.of(String) with enhanced error handling

This update includes validation for missing disclosures, duplicate disclosure digests, and malformed disclosure data, improving overall robustness and error handling during disclosure processing.

Closes #33020

Signed-off-by: Ogenbertrand <ogenbertrand@gmail.com>

francis-pouatcha commented:
@ahus1 is there anything missing to merge this pull request?

ahus1 reviewed (commented):
Thank you for the PR. See below for a something that I think is a mismatch of the check and the error message. 

Not sure if you want to change the check, or the error message. As I'm not into SDJ-JWT , I'm marking it as a comment and leave it to you to find the answer.

ahus1 commented:
@francis-pouatcha / @IngridPuppet - please review the latest changes if they look good to you. Thanks!

IngridPuppet commented on core/src/main/java/org/keycloak/sdjwt/vp/SdJwtVP.java:
```diff
@@ -104,8 +104,16 @@ public static SdJwtVP of(String sdJwtString) {
+
         String issuerSignedJWTString = sdJwtString.substring(0, disclosureStart);
-        String disclosuresString = sdJwtString.substring(disclosureStart + 1, disclosureEnd);
+        String disclosuresString ="";
```
Better with spaces around the equal sign.
```suggestion
        String disclosuresString = "";
```

IngridPuppet commented on core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPTest.java:
```diff
@@ -196,4 +198,24 @@ public void testS6_2_PresentationPartialDisclosure() throws VerificationExceptio
+        assertEquals(4, sdJwtVP.getDisclosures().size());
+    }
+
+    @Test
+    public void testOf_MalformedSdJwt_ThrowsIllegalArgumentException() {
+        // Given
+        String malformedSdJwt = "issuer-signed-jwt"; // missing delimiter at the end
+
+        // When & Then
+        var exception = assertThrows(IllegalArgumentException.class, () -> SdJwtVP.of(malformedSdJwt));
+        assertEquals("SD-JWT is malformed, expected to end with ~", exception.getMessage());
+    }
```
This test will likely fail as is due to the asserted error message not aligning with the changes. Additionally, the test vector should have a different comment, since as highlighted by  @ahus1, the problem is not that the SD-JWT is not ending with ~.

Ogenbertrand replied:
@IngridPuppet Thank you for pointing out the issue with the test. I'll update the test to reflect the changes made to the error message. I'll make sure to address this in the next update.

IngridPuppet reviewed (changes requested):
Thank you @ahus1 for the extra attention, and thank you @Ogenbertrand for addressing the comments. I have some minor additional concerns.

Ogenbertrand committed:
Fix SD-JWT parsing to correctly handle delimiter

Updated the SD-JWT parsing logic to accurately check for the presence of the delimiter character. The exception message has been corrected to reflect the actual check being performed.

Update SdJwtVPTest to use explicit type instead of var keyword

Closes #33020

Signed-off-by: Ogenbertrand <ogenbertrand@gmail.com>

Ogenbertrand committed:
Merge branch 'main' into issuer-#33020

ahus1 commented:
@Ogenbertrand - please avoid rebasing this PR or merging the main branch into this PR as it will reset the CI jobs every time. This is a waste of CI resources, and it will also reset any previous review. 

Don't rebase a PR unless there are conflicts or a maintainer asks you to rebase the PR.

Ogenbertrand commented:
> @Ogenbertrand - please avoid rebasing this PR or merging the main branch into this PR as it will reset the CI jobs every time. This is a waste of CI resources, and it will also reset any previous review.
> 
> Don't rebase a PR unless there are conflicts or a maintainer asks you to rebase the PR.

Okay @ahus1, Understood, i was actually trying to fix the error with the previous run.

ahus1 reviewed (approved):
Approving based on previews reviews. Thanks!
